rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,401 @@
+```
+- Unpin a notification.
+```
+POST /api/notifications/unpin
+{ notification_id: string[] }
+```
+
+Writing to the same Notification storage might cause conflict on the `write`. If the storage mechanism implementation
+doesn’t make them impossible, we will have to use a single place to change notification status (via the [queue](#localrepository)).
+
+## Settings
+Provides HTTP API for Kibana UI to configure user-specific settings.
+Not provided in MVP, we might add in the future:
+- Unsubscribe from a source
+- Subscribe to a source
+
+#### Filtering
+Filtering allows users to hide notifications they aren’t interested in. Kibana UI attaches a list of filters when requesting notifications.
+Kibana stores filter as a part of User settings data to provide a seamless experience between different deployments on Cloud and user sessions.
+
+Filter configuration is passed to `Repository.get` method as `filter` property.
+```typescript
+interface NotificationFilter {
+  source_type?: string[];
+  group_id?: string[];
+  read?: boolean;
+}
+```
+
+# Drawbacks
+
+- It is not entirely clear what functionality will be supported by UNS and when this will be known.
+It forces us to evaluate each feature separately as to how difficult it will be to adapt it to UNS later.
+- The service relies on the yet-to-be-added ability to uniquely identify the user, enumerate users with a given set of permissions.
+- The service doesn't provide a built-in security mechanism but relies on a notification source implementation.
+- Lack of a centralized i18n service makes message transferring over-bloated due to the necessity to send all the possible translations over the wire.
+
+
+# Alternatives
+
+### Treat Cloud as another pluggable storage mechanism.
+Kibana notification service collects internal notifications and shows them in the UI for both on-prem and Cloud environments.
+Kibana reuses Notification service implementation on Cloud but enhances it with additional pluggable storage of external notifications.
+![img](../images/0014/alternative_implementation.png)
+
+The main benefit of this approach is that Kibana always uses the same service in any environment.
+But in addition to the restrictions listed in the [Drawbacks](#drawbacks) section, the following are added:
+- Each retrieval of a notification list still requires waiting for a roundtrip to UNS.
+- The list of Kibana notifications stored in an instance might be out of sync with a copy stored in UNS due to network problems.
+- Storing notifications in the Cloud deployment and their copy in UNS consumes twice the disk space.
+
+# Adoption strategy
+
+We start with introducing Kibana Notification Service as a Kibana plugin, adding more notification sources along the road.
+Integration with UNS is something that is not going to happen any time soon. The work can be divided into the following phases:
+
+### Phase I
+Kibana shows notifications created by the Newsfeed Kibana plugin in Notification flyout.
```

Comment:
Do you know how we're going to "pull" from the newsfeed and persist these notifications locally? Will this be using TaskManager? Are there any situations where we currently delete newsfeed items that complicate this implementation?

Reply:
> Are there any situations where we currently delete newsfeed items that complicate this implementation?

The current Newsfeed implementation doesn't store any data. It fetches notifications from the browser on every visit and shows them in UI.
On `Phase I` we start with moving the existing Newsfeed plugin logic to the Notification service and add transformation layer to convert https://github.com/elastic/newsfeeds/blob/master/data/kibana.yaml#L10-L17 to the notification compatible interface. News is shown in Notification UI, but not stored in Notification storage.

> Do you know how we're going to "pull" from the newsfeed and persist these notifications locally? Will this be using TaskManager? Are there any situations where we currently delete newsfeed items that complicate this implementation?

Since `task_manager` cannot be used outside of the `x-pack` folder, we will have to implement a lock mechanism via ES to allow only a single Kibana instance to retrieve the news and periodically check whether updates are available. 

Newsfeed plugin can fetch notifications and create a notification for every news item. Every news has `hash` property, which seems to be stable for different versions, see [7.11.0](https://feeds-staging.elastic.co/kibana/v7.11.0.json), [7.10.0](https://feeds-staging.elastic.co/kibana/v7.10.0.json). We can make notification creation idempotent operation and use `hash` as `id`  to avoid creating duplicates:
```
{
  "script": {
    "source": "if (ctx._source.id = params.notification.id) { ctx.op = 'none' }",
...
```

Or we have to track manually what news already created notifications.